subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -266,7 +268,59 @@ static int hfp_ag_next_step(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *us
 
 static int hfp_ag_send(struct bt_hfp_ag *ag, struct bt_ag_tx *tx)
 {
-	return bt_rfcomm_dlc_send(&ag->rfcomm_dlc, tx->buf);
+	int err = bt_rfcomm_dlc_send(&ag->rfcomm_dlc, tx->buf);
+
+	if (err < 0) {
+		net_buf_unref(tx->buf);
+	}
+
+	return err;
+}
+
+static void bt_ag_tx_work(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct bt_hfp_ag *ag = CONTAINER_OF(dwork, struct bt_hfp_ag, tx_work);
+	sys_snode_t *node;
+	struct bt_ag_tx *tx;
+	bt_hfp_state_t state;
+
+	hfp_ag_lock(ag);
+	state = ag->state;
+	hfp_ag_unlock(ag);
+	if ((state == BT_HFP_DISCONNECTED) || (state == BT_HFP_DISCONNECTING)) {
+		LOG_ERR("AG %p is not connected", ag);
+		return;
+	}
+
+	hfp_ag_lock(ag);
+	node = sys_slist_peek_head(&ag->tx_pending);
+	if (!node) {
+		hfp_ag_unlock(ag);
+		LOG_DBG("No pending tx");
+		return;
+	}
```

Comment:
Why do you have a different pattern for locking and unlocking with the first and second branch? In the first one you lock before the branch and in the second one in the branch.

Generally, when you have a function which holds a lock for a large part of the function, the typical pattern is:

```c
static void my_func(struct bt_hfp_ag *ag)
{
	int err;

	hfp_ag_lock(ag);
	err = foo(ag);
	if (err) {
		LOG_ERR("...");
		goto unlock;
	}	

	err = bar(ag);
	if (err) {
		LOG_ERR("...");
		goto unlock;
	}	

	...

unlock:
	hfp_ag_unlock(ag);
	return err;
}
```

The exception to the above is if the function does callback or anything that might introduce re-entrancy to try to acquire the lock a second time. In such cases you need to release and re-acquire the lock, or use a different lock type which permits the same thread to take the lock multiple times.

Reply:
Updated